File: remote/src/main/java/com/kelsos/mbrc/presenters/GenreArtistsPresenterImpl.java
package com.kelsos.mbrc.presenters;

import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.MetaDataType;
import com.kelsos.mbrc.annotations.Queue;
import com.kelsos.mbrc.domain.Artist;
import com.kelsos.mbrc.interactors.QueueInteractor;
import com.kelsos.mbrc.interactors.library.GenreArtistInteractor;
import com.kelsos.mbrc.ui.views.GenreArtistView;
import timber.log.Timber;

public class GenreArtistsPresenterImpl implements GenreArtistsPresenter {
  @Inject private GenreArtistInteractor genreArtistInteractor;
  @Inject private QueueInteractor interactor;

  private GenreArtistView view;

  @Override public void bind(GenreArtistView view) {
    this.view = view;
  }

  @Override public void onPause() {

  }

  @Override public void onResume() {

  }

  @Override public void load(long genreId) {
    genreArtistInteractor.getGenreArtists(genreId).subscribe(artists -> {
      view.update(artists);
    }, t -> {
      Timber.v(t, "Failed");
    });
  }

  @Override public void queue(@Queue.Action String action, Artist artist) {
    interactor.execute(MetaDataType.ARTIST, action, (int) artist.getId()).subscribe(isSuccessful -> {
      if (isSuccessful) {
        view.onQueueSuccess();
      } else {
        view.onQueueFailure();
      }
    }, throwable -> {
      Timber.e(throwable, "Queueing failed");
      view.onQueueFailure();
    });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/presenters/GenreArtistsPresenterImpl.kt
package com.kelsos.mbrc.presenters

import com.kelsos.mbrc.annotations.MetaDataType
import com.kelsos.mbrc.annotations.Queue
import com.kelsos.mbrc.domain.Artist
import com.kelsos.mbrc.interactors.QueueInteractor
import com.kelsos.mbrc.interactors.library.GenreArtistInteractor
import com.kelsos.mbrc.ui.views.GenreArtistView
import timber.log.Timber
import javax.inject.Inject

class GenreArtistsPresenterImpl
@Inject constructor(private val genreArtistInteractor: GenreArtistInteractor,
                    private val interactor: QueueInteractor) : GenreArtistsPresenter {

  private var view: GenreArtistView? = null

  override fun bind(view: GenreArtistView) {
    this.view = view
  }

  override fun onPause() {

  }

  override fun onResume() {

  }

  override fun load(genreId: Long) {
    genreArtistInteractor.getGenreArtists(genreId).subscribe({ view!!.update(it) }) {
      Timber.v(it, "Failed")
    }
  }

  override fun queue(@Queue.Action action: String, artist: Artist) {
    interactor.execute(MetaDataType.ARTIST, action, artist.id).subscribe({
      if (it) {
        view?.onQueueSuccess()
      } else {
        view?.onQueueFailure()
      }
    }) {
      Timber.e(it, "Queueing failed")
      view?.onQueueFailure()
    }
  }
}
